src/transport/SecureSessionTable.cpp

```diff
@@ -0,0 +1,277 @@
+    // an existing session in the table in normal operating circumstances.
+    //
+    if (mEntries.Allocated() < CHIP_CONFIG_SECURE_SESSION_POOL_SIZE)
+    {
+        allocated = mEntries.CreateObject(*this, secureSessionType, sessionId.Value());
+        VerifyOrExit(allocated != nullptr, rv = Optional<SessionHandle>::Missing());
+    }
+    else
+    {
+        allocated = EvictAndAllocate(sessionId.Value(), secureSessionType, sessionEvictionHint);
+    }
+
+    rv             = MakeOptional<SessionHandle>(*allocated);
+    mNextSessionId = sessionId.Value() == kMaxSessionID ? static_cast<uint16_t>(kUnsecuredSessionId + 1)
+                                                        : static_cast<uint16_t>(sessionId.Value() + 1);
+
+exit:
+    return rv;
+}
+
+SecureSession * SecureSessionTable::EvictAndAllocate(uint16_t localSessionId, SecureSession::Type secureSessionType,
+                                                     const ScopedNodeId & sessionEvictionHint)
+{
+    VerifyOrDieWithMsg(!mRunningEvictionLogic, SecureChannel,
+                       "EvictAndAllocate isn't re-entrant, yet someone called us while we're already running");
+    mRunningEvictionLogic = true;
+
+    ChipLogProgress(SecureChannel, "Evicting a slot for session with LSID: %d, type: %u", localSessionId,
+                    (uint8_t) secureSessionType);
+
+    VerifyOrDie(mEntries.Allocated() <= CHIP_CONFIG_SECURE_SESSION_POOL_SIZE);
+
+    //
+    // Create a temporary list of objects each of which points to a session in the existing
+    // session table, but are swappable. This allows them to then be used with a sorting algorithm
+    // without affecting the sessions in the table itself.
+    //
+    Platform::ScopedMemoryBufferWithSize<SortableSession> sortableSessions;
+    sortableSessions.Calloc(mEntries.Allocated());
+    if (!sortableSessions)
+    {
+        VerifyOrDieWithMsg(true, SecureChannel, "We couldn't allocate a session!");
+        return nullptr;
+    }
+
+    int index = 0;
+    ForEachSession([&index, &sortableSessions](auto session) {
+        sortableSessions.Get()[index].mSession = session;
+        index++;
+        return Loop::Continue;
+    });
+
+    auto sortableSessionSpan = Span<SortableSession>(sortableSessions.Get(), mEntries.Allocated());
+    EvictionPolicyContext policyContext(sortableSessionSpan, sessionEvictionHint);
+
+    DefaultEvictionPolicy(policyContext);
+    ChipLogProgress(SecureChannel, "Sorted sessions for eviction...");
+
+    auto numSessions = mEntries.Allocated();
+
+#if CHIP_DETAIL_LOGGING
+    ChipLogDetail(SecureChannel, "Sorted Eviction Candidates (ranked from best candidate to worst):");
+    auto i = 0;
+    for (auto * session = sortableSessions.Get(); session != (sortableSessions.Get() + numSessions); session++, i++)
+    {
+        ChipLogDetail(SecureChannel, "\t%d: [%p] -- State: '%s', ActivityTime: %lu", i, session->mSession,
```

Comment:
Do you need `i` at all?  Can't you use `session - sortableSessions.Get()` for that?